File: src/mandelbrot.frontend/Startup.fs
namespace mandelbrot.frontend

open System
open Microsoft.AspNetCore
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open WebSharper.AspNetCore
open asplib.Remoting
open asplib.Services
open iselenium

type Startup(env: IWebHostEnvironment, configuration: IConfiguration) =
    do RemotingContext.Set(env, configuration) |> ignore

    new(configuration: IConfiguration) = // ServiceProvider constructor for tests
        Startup(null, configuration)

    member this.ConfigureServices(services: IServiceCollection) =
        services.AddSingleton(configuration)
            .AddDistributedMemoryCache() // session
            .AddImageCache()    // by Image.Coordinates
            .AddMemoryCache()   // by Image.Gid
            .AddImageService()  // Singleton for static access
            .AddSingleton<IWebSocketClient>(fun _ ->
                new ServerWebSocketClient(new Uri(configuration.GetValue<string>("RendererHost") +
                                            mandelbrot.renderer.Site.RendererWebSocket))
                :> IWebSocketClient
            )
            .AddHttpClient()    // IHttpClientFactory for GET image
            .AddSession(fun options ->
                options.IdleTimeout <- TimeSpan.FromMinutes(30.0)
                options.Cookie.HttpOnly <- true
            )
            .AddSitelet(Site.Main)
            .AddLogging()
        |> ignore

    member this.OnShutDown(disposable: IDisposable) =
        disposable.Dispose()

    member this.Configure(app: IApplicationBuilder,
                            env: IWebHostEnvironment,
                            applifetime: IHostApplicationLifetime,
                            disposable: ImageService) =
        if (env.IsDevelopment()) then
            WebSharper.Web.Remoting.DisableCsrfProtection() |> ignore
            app.UseDeveloperExceptionPage()
                .UseNoCache() |> ignore
            StaticServiceProvider.SetProvider(app.ApplicationServices);
        else
            app.UseHttpsRedirection()
                .UseHsts() |> ignore

        applifetime.ApplicationStopping.Register(fun _ -> this.OnShutDown(disposable)) |> ignore

        app.UseISelenium()
            .UseImageSocket(Site.FrontendWebSocket)
            .UseDefaultFiles()  // [<SPAEntryPoint>]
            .UseStaticFiles()   // [<SPAEntryPoint>]
            .UseSession()
            .UseRequestQuerySession()
            .UseWebSharper()
            .Run(fun context ->
                context.Response.StatusCode <- 404
                context.Response.WriteAsync("Page not found"))

module Program =
    let CreateHostBuilder args =
        Host.CreateDefaultBuilder(args)
            .ConfigureWebHostDefaults(fun webBuilder ->
                webBuilder
                    .UseStartup<Startup>()
                    .CaptureStartupErrors(true)
                    .ConfigureLogging(fun logging ->
                        logging.AddConsole()
                                .AddDebug() |> ignore
                    )
                    .UseSetting("detailedErrors", "true") |> ignore
            )

    [<EntryPoint>]
    let main args =
        CreateHostBuilder(args).Build().Run()
        0

File: src/minimal.websharper.spa/Startup.cs
﻿using asplib.Model.Db;
using asplib.Remoting;
using iselenium;
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using System;
using WebSharper.AspNetCore;

namespace minimal.websharper.spa
{
    public class Startup
    {
        public IConfigurationRoot Configuration { get; }
        public IWebHostEnvironment Environment { get; }
        public static HttpContext HttpContext { get; set; }

        public Startup(IWebHostEnvironment env)
        {
            Environment = env;
            var dom = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddEnvironmentVariables()
                .Build();
            Configuration = dom;
            RemotingContext.Set(Environment, Configuration);
            ASP_DBEntities.ConnectionString = Configuration["ASP_DBEntities"];  // globally
        }

        public void ConfigureServices(IServiceCollection services)
        {
            // WebSharper empty
            services.AddDistributedMemoryCache()
                .AddSession(options =>
                {
                    options.IdleTimeout = TimeSpan.FromMinutes(30);
                    options.Cookie.HttpOnly = true;
                })
                .AddLogging()
                .AddSitelet(TestResultSite.Main);
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            //WebSharper.Web.Remoting.DisableCsrfProtection();    // HTTP 403 prevention not needed here
            app.UseDeveloperExceptionPage()
                .UseISelenium()
                .UseNoCache()
                .UseDefaultFiles()
                .UseSession()
                .UseRequestQuerySession()
                .UseStaticFiles()
                .UseWebSharper()
                .Run(context =>
                {
                    HttpContext = context;
                    context.Response.StatusCode = 404;
                    return context.Response.WriteAsync("Page not found");
                });
        }

        public static void Main(string[] args)
        {
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .CaptureStartupErrors(true)
                .UseSetting("detailedErrors", "true")
                .Build()
                .Run();
        }
    }
}